### ISSUE TITLE
False positive Magic Number violations within the method declaration in class field

### ISSUE THREAD

**Issue by Anmol202005** [checkstyle/checkstyle#16228]
I have read check documentation: https://checkstyle.org/checks/coding/magicnumber.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

Detected at: https://github.com/checkstyle/checkstyle/pull/16172#discussion_r1927955827

``` 
$ javac Test.java 
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="MagicNumber">
            <property name="ignoreNumbers" value="-2,-1,0,1,2,100"/>

            <property name="ignoreFieldDeclaration" value="true"/>

            <property name="constantWaiverParentToken" value="ARRAY_INIT, ASSIGN, ELIST, EXPR"/>
        </module>
    </module>
</module>

$ cat Test.java 
import java.util.concurrent.Callable;
public class Test {

    private final Callable<Void> SLEEP_FOR_A_DAY_NS = () -> {
        Thread.sleep(86400_000);
        return null;
    };

    private final Callable<Void> SLEEP_FOR_A_DAY_EXP_NS = new Callable<Void>() {
        @Override
        public Void call() throws InterruptedException {
            Thread.sleep(86400_000); // violation 'False positive'
            return null;
        }
    };
}

$ java -jar checkstyle-10.21.1-all.jar -c config.xml  Test.java
Starting audit...
[ERROR] /var/tmp/Test.java:5:22: '86400_000' is a magic number. [MagicNumber]
[ERROR] /var/tmp/Test.java:12:26: '86400_000' is a magic number. [MagicNumber]
Audit done.
Checkstyle ends with 2 errors.
```

Recommended Investigation:
- Examine how anonymous classes are parsed.
- Review the current magic number detection implementation

This issue will be a follow up of PR: #16172

Expected Outcome:
No violations within the method declaration in SLEEP_FOR_A_DAY_EXP_NS

**Comment by Anmol202005** [comment]
> [@Anmol202005](https://github.com/Anmol202005) please strictly follow the issue template with a jar built from your PR branch.
done.

**Comment by nrmancuso** [comment]
These two fields are essentially the same code just with different syntax. Regardless, if we are ignoring fields I feel that there should be no violations at all on/in the anonymous class because it is a field declaration. @rnveach what do you think? Please approve if you agree.

**Comment by rnveach** [comment]
Are we sure this isn't a duplicate of https://github.com/checkstyle/checkstyle/issues/16101 or any other MagicNumber issue?

I assume we are saying this is an issue because of `ignoreFieldDeclaration`.

**Comment by romani** [comment]
No violations are expected.
Duplicate issue are natural processes, eventually cleaned up. Usually most fresh issues re resolved sooner :).

**Comment by romani** [comment]
@Anmol202005 , if you have time please help to fix issue, you may already well experienced in this Check implementation.

**Comment by Anmol202005** [comment]
@romani 

'''
class C2 {
    Person annon = new Person(55);
    Object anon = new Object() {
        public void magicMethod() {
            Set<Person> p1 = new HashSet<Person>(5); // violation ''5' is a magic number'
            Object[] values = new Object[3]; // violation ''3' is a magic number'
            String[] fStrings = new String[3]; // violation ''3' is a magic number'
            Person p12 = new Person(8); // violation ''8' is a magic number'

            double double_var1 = 0D;
            double_var1 = 1.0 + 3.0; // violation ''3.0' is a magic number'
            double double_magic1 = 1.5_0; // violation ''1.5_0' is a magic number'
            if (1.0 < 3.0); // violation ''3.0' is a magic number'
            double_magic1 *= 1.5; // violation ''1.5' is a magic number'
        }
    };
}
'''

are all these violations valid !!
kindly clarify .

**Comment by Anmol202005** [comment]
@romani 

```
class C2 {
    Person annon = new Person(55);
    Object anon = new Object() {
        public void magicMethod() {
            Set<Person> p1 = new HashSet<Person>(5); // violation ''5' is a magic number'
            Object[] values = new Object[3]; // violation ''3' is a magic number'
            String[] fStrings = new String[3]; // violation ''3' is a magic number'
            Person p12 = new Person(8); // violation ''8' is a magic number'

            double double_var1 = 0D;
            double_var1 = 1.0 + 3.0; // violation ''3.0' is a magic number'
            double double_magic1 = 1.5_0; // violation ''1.5_0' is a magic number'
            if (1.0 < 3.0); // violation ''3.0' is a magic number'
            double_magic1 *= 1.5; // violation ''1.5' is a magic number'
        }
    };
}
```
Are all these violations valid !

kindly clarify .

**Comment by Anmol202005** [comment]
@romani 
It might be confusing to the users how the current Magic Number check works with IgnoreFieldDeclaration = true even for Lambda functions.

```
private final Callable<Void> SLEEP_FOR_A_DAY_NS = () -> {
        Thread.sleep(86400_000);
        return null;
    };
    private final BiFunction<Integer, Integer, Integer> ADD_AND_SQUARE_NS = (a, b) -> {
        int sum = a + b + 5;  // violation ''5' is a magic number'
        sum = 5;  //ok for this.
        return sum * sum * 69;
    };

```
current checks shows a violation for variable declaration , but doesn't show violation for variable assignment.

**Comment by Anmol202005** [comment]
@romani 
I do agree but it seem ambiguous how variable assignment throws violation here for `IgnoreFieldDeclaration = true` 

```
class C2 {
    Person annon = new Person(55);
    Object anon = new Object() {
        public void magicMethod() {
            Set<Person> p1 = new HashSet<Person>(5); // violation ''5' is a magic number'
            Object[] values = new Object[3]; // violation ''3' is a magic number'
            String[] fStrings = new String[3]; // violation ''3' is a magic number'
            Person p12 = new Person(8); // violation ''8' is a magic number'

            double double_var1 = 0D;
            double_var1 = 1.0 + 3.0; // violation ''3.0' is a magic number'
            double double_magic1 = 1.5_0; // violation ''1.5_0' is a magic number'
            if (1.0 < 3.0); // violation ''3.0' is a magic number'
            double_magic1 *= 1.5; // violation ''1.5' is a magic number'
        }
    };
}
```

but it does not throw any violation for lambda functions, there is no consistency !!

In my opinion either we need to report violations even in lambda functions or should ignore them completely.

**Comment by romani** [comment]
Please make same example with lambda and share output in CLI format, to be absolutely clear.

**Comment by Anmol202005** [comment]
```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="MagicNumber">
            <property name="ignoreNumbers" value="-2,-1,0,1,2,100"/>

            <property name="ignoreFieldDeclaration" value="true"/>

            <property name="constantWaiverParentToken" value="ARRAY_INIT, ASSIGN, ELIST, EXPR"/>
        </module>
    </module>
</module>
```

Lambda function : 

```
package com.puppycrawl.tools.checkstyle.checks.coding.magicnumber;

import java.util.HashSet;
import java.util.Set;

public class InputMagicNumberTesty {

    private Runnable magicMethodLambda = () -> {
    Set<Person> p1 = new HashSet<Person>(5);
    Object[] values = new Object[3];
    String[] fStrings = new String[3];
    Person p12 = new Person(8);

    double double_var1 = 0D;
    double_var1 = 1.0 + 3.0;
    double double_magic1 = 1.5_0;
    if (1.0 < 3.0);
    double_magic1 *= 1.5;
};
}
```

```
PS C:\Users\anmol\IdeaProjects\checkstyle3\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\magicnumber> java -jar .\checkstyle-10.21.1-all.jar -c .\magic_config.xml .\InputMagicNumberTesty.java
Starting audit...
[ERROR] C:\Users\anmol\IdeaProjects\checkstyle3\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\magicnumber\.\InputMagicNumberTesty.java:10:42: '5' is a magic number. [MagicNumber]
[ERROR] C:\Users\anmol\IdeaProjects\checkstyle3\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\magicnumber\.\InputMagicNumberTesty.java:11:34: '3' is a magic number. [MagicNumber]
[ERROR] C:\Users\anmol\IdeaProjects\checkstyle3\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\magicnumber\.\InputMagicNumberTesty.java:12:36: '3' is a magic number. [MagicNumber]
[ERROR] C:\Users\anmol\IdeaProjects\checkstyle3\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\magicnumber\.\InputMagicNumberTesty.java:13:29: '8' is a magic number. [MagicNumber]
[ERROR] C:\Users\anmol\IdeaProjects\checkstyle3\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\magicnumber\.\InputMagicNumberTesty.java:17:28: '1.5_0' is a magic number. [MagicNumber]
Audit done.
Checkstyle ends with 5 errors.

```

**Comment by romani** [comment]
@Anmol202005 , thanks a lot for your time on this issue.

I am closing issue, we should not fix anything.
